aGHQ.qmd

```diff
@@ -0,0 +1,1402 @@
+
+```{julia}
+#| output: false
+function StatsAPI.fit!(m::BernoulliPIRLS)
+  θβ = m.θβ
+  pp1 = length(θβ)                # length(β) = p and length(θ) = 1
+  opt = Opt(:LN_BOBYQA, pp1)
+  mβ = view(θβ, 2:pp1)
+  function obj(x, g)
+    if !isempty(g)
+      throw(ArgumentError("gradient not provided, g must be empty"))
+    end
+    copyto!(θβ, x)
+    mul!(m.ytbl.offset, m.X, mβ)
+    return laplaceapprox(pirls!(m))
+  end
+  opt.min_objective = obj
+  lb = fill!(similar(θβ), -Inf)   # vector of lower bounds
+  lb[1] = 0                       # scalar θ must be non-negative
+  NLopt.lower_bounds!(opt, lb)
+  minf, minx, ret = optimize(opt, copy(θβ))
+  @info (; ret, fevals=opt.numevals, minf)
+  return m
+end
+```
+
+```{julia}
+fit!(m);
+```
+
+```{julia}
+#| code-fold: true
+println("Converged to θ = ", first(m.θβ), " and β =")
+showcompact(view(m.θβ, 2:lastindex(m.θβ)))
+```
+
+These estimates differ somewhat from those for model `com05`.
+
+```{julia}
+#| code-fold: true
+println(
+  "Estimates for com05: θ = ",
+  only(com05.θ),
+  ", fmin = ",
+  deviance(com05),
+  ", and β =",
+)
+showcompact(com05.β)
+```
+
+The discrepancy in the results is because the `com05` results are based on a more accurate approximation to the integral called *adaptive Gauss-Hermite Quadrature*, which is discussed in @sec-aGHQ.
+
+### Generalizations to more complex structure
+
+There is an implicit assumption in the `BernoulliPIRLS` structure that random effects in the model are simple, scalar random effects associated with a single grouping factor, which is represented by `m.ytbl.refs`.
+For such models the random effects model matrix, $\bbZ$, is an $n\times q$ indicator matrix, the covariance parameter, $\bbtheta$, is one-dimensional and the covariance factor, $\bbLambda_{\bbtheta}=\theta\,\bbI_q$ is a scalar multiple of the $q\times q$ identity matrix, $\bbI_q$.
+Furthermore, $\bbLambda_{\bbtheta}'\bbZ'\bbW\bbZ\bbLambda_{\bbtheta}+\bbI_q$ is also diagonal, as is its Cholesky factor, $\bbL$.
+
+We have taken advantage of the special structure of these matrices both in representations --- storing $\bbL$ by storing only the diagonal values in the vector `m.utbl.Ldiag` --- and in some algorithms for the PIRLS iterative step.
+
+The PIRLS algorithm to determine the conditional mode, $\tilde{\bbu}$, of the random effects, $\mcU$, and Laplace's approximation to the log-likelihood for GLMMs can be generalized to models with vector-valued random effects, or with random effects associated with more than one grouping factor, or with both.
+The more general representation of $\bbZ$ and $\bbL$ used with linear mixed models can be adapted for GLMMs as well.
+
+We chose to specialize the representation of GLMMs in this appendix to this specific type of random effects to be able to demonstrate adaptive Gauss-Hermite quadrature in @sec-aGHQ, which, at present, is restricted to models with a single, simple, scalar, random effects term.
```

Comment:
Do we throw an error for vector valued RE and nAGQ > 1?

Reply:
I think we just ignore an `nAGQ > 1` unless we have a single scalar random effects term.  We could add a warning or info message that it will be ignored, in that case.

I do have a paper on multivariate Gauss-Hermite quadrature but have only glanced at it.  The folks who wrote the multilevel modeling for Stata, Sophia Rabe-Hesketh and Anders Skrondal, claim that they are using Gauss-Hermite quadrature for vector-valued random effects but I don't know the details.